src/cycleclock.h

```diff
@@ -167,16 +175,22 @@ inline BENCHMARK_ALWAYS_INLINE int64_t Now() {
 #elif defined(__riscv) // RISC-V
   // Use RDCYCLE (and RDCYCLEH on riscv32)
 #if __riscv_xlen == 32
-  uint64_t cycles_low, cycles_hi0, cycles_hi1;
-  asm("rdcycleh %0" : "=r"(cycles_hi0));
-  asm("rdcycle %0" : "=r"(cycles_lo));
-  asm("rdcycleh %0" : "=r"(cycles_hi1));
-  // This matches the PowerPC overflow detection, above
-  cycles_lo &= -static_cast<int64_t>(cycles_hi0 == cycles_hi1);
-  return (cycles_hi1 << 32) | cycles_lo;
+  uint32_t cycles_lo, cycles_hi0, cycles_hi1;
+  // This asm also includes the PowerPC overflow handling strategy, as above.
+  // Implemented in assembly because Clang insisted on branching.
```

Comment:
I really wonder if this is clang risc-v backend being too aggressive in lowering that into branches.

Reply:
It really depends on your scheduling model, assumptions on which branch is more likely, etc. But I guess from the point of view of the compiler it's like "what do you mean I shouldn't branch? I didn't write anything anywhere to even suggest that I shouldn't do that!" :)
Curiously, it's actually more typical to be the other way around, since LLVM's IR more natively relies on the notion of conditional moves, while GCC seems to be more branch oriented.

Reply:
Yeah, it is indeed a balance between "what's the branch misprediction cost" vs "is this branch predictable" vs "is this actually perf-sensitive" vs "how much more instrs are we willing to do to avoid the branch"
The other thing is that as far as i understand for RISC-V, conditional move instruction
will only be available in `B`it manipulation ISA, and i guess currently such `select`
is simply expanded into a branch instead of some branch-less and-of-sub sequence.

Reply:
Yup, that's exactly right. In fact, one of the first things I worked on in LLVM's RISC-V backend was improving the code generation when you `select`ed a value that had to be moved in two steps, as LLVM would mangle that up into a less than ideal branch lowering mess.